#### &#128337; Temps previst de dedicació: 20 minuts

# TEMA 4: Treball amb API

> <br>🚨&nbsp; **Atenció**
> 
> <span style="color: #f69;">En aquest nivell B2 explorem conceptes avançats de JavaScript que excedeixen les capacitats d'un notebook. Descarregueu [Visual Studio Code](https://code.visualstudio.com) a la vostra màquina i creeu projectes locals per a cada exemple proposat, incloent els fitxers HTML, CSS i JavaScript que calguin en cada cas. Instal·leu també les extensions [Live Server](https://marketplace.visualstudio.com/items?itemName=ritwickdey.LiveServer) i [Thunder Client](https://marketplace.visualstudio.com/items?itemName=rangav.vscode-thunder-client). La primera us permetrà executar el vostre codi en un servidor web local des de Visual Studio Code. La segona us resultarà especialment útil per poder treballar més còmodament amb diferents API a mida que estudieu les unitats d'aquest Tema 4.</span>
> <br><br>

## Unitat 2: Peticions GET i POST amb Fetch

En el nivell B1 vàrem tenir un primer contacte amb les peticions GET amb Fetch. En concret, vàrem exposar el que diu a continuació:

_El 2015 [...] JavaScript va incorporar de manera nativa un nou mecanisme, l'API Fetch, pensat per succeir `XMLHttpRequest` i incorporant mecanismes com les promeses, [...] que faciliten la petició de dades sense haver d'utilitzar funcions _callback_. Actualment, l'API Fetch s'ha convertit en l'estàndard _de facto_ per fer peticions de manera asíncrona._ 

_Fetch és una API pròpia de JavaScript (per tant, una part del llenguatge que podem utilitzar directament) que ens permet portar a terme el mateix tipus d'operacions que el mecanisme AJAX descrit a la unitat anterior, però d'una manera més senzilla i amb una sintaxi més fàcil de comprendre; està aconseguint que cada vegada més programadors la utilitzin en detriment de `XMLHttpRequest`._ 

I ho vam poder veure amb un exemple molt bàsic de petició `GET` utilitzant l'API de Star Wars:

In [1]:
async function logStarWarsPeople() {
  let response = await fetch('https://swapi.dev/api/people');
  let data = await response.text();
  console.log(data);
}

logStarWarsPeople();

Promise { <pending> }

{"count":82,"next":"https://swapi.dev/api/people/?page=2","previous":null,"results":[{"name":"Luke Skywalker","height":"172","mass":"77","hair_color":"blond","skin_color":"fair","eye_color":"blue","birth_year":"19BBY","gender":"male","homeworld":"https://swapi.dev/api/planets/1/","films":["https://swapi.dev/api/films/1/","https://swapi.dev/api/films/2/","https://swapi.dev/api/films/3/","https://swapi.dev/api/films/6/"],"species":[],"vehicles":["https://swapi.dev/api/vehicles/14/","https://swapi.dev/api/vehicles/30/"],"starships":["https://swapi.dev/api/starships/12/","https://swapi.dev/api/starships/22/"],"created":"2014-12-09T13:50:51.644000Z","edited":"2014-12-20T21:17:56.891000Z","url":"https://swapi.dev/api/people/1/"},{"name":"C-3PO","height":"167","mass":"75","hair_color":"n/a","skin_color":"gold","eye_color":"yellow","birth_year":"112BBY","gender":"n/a","homeworld":"https://swapi.dev/api/planets/1/","films":["https://swapi.dev/api/films/1/","https://swapi.dev/api/films/2/","http

Recordeu que després de mostrar aquest exemple vam comentar el que diu a continuació:

_És important notar la presència de les paraules clau `await` i `async`: la funció `fetch()` és una funció asíncrona que no retorna directament una resposta, sinó que retorna només una garantia... una promesa (és a dir, un objecte de tipus `Promise`) que més endavant es rebrà una resposta. [...] `await` és una paraula clau que indica que l'execució del codi es deté fins que la promesa es materialitza en una resposta._

_La resposta que, tot i que no de manera immediata, ens acabarà retornant la funció `fetch()` és d'un tipus específic `Response` que té algunes propietats d'ús molt freqüent:_

- _`status` ens indica el codi numèric amb el resultat de la petició [...]_
- _`text()` [retorna] una promesa [...] que es materialitzarà en les dades sol·licitades al servidor en format text._
- _`json()` [retorna] una promesa que es materialitzarà en les dades sol·licitades al servidor en format JSON._
  
Com a introducció bàsica al funcionament de Fetch, en el nivell B1 ens vam quedar amb la idea que la funció `fetch()` ens permet fer peticions asíncrones de manera més senzilla que amb `XMLHttpRequest`. 

En aquest nivell B2 aprofundirem en el funcionament de Fetch. En realitat, quan operem amb una API podem no només fer peticions `GET` per llegir dades, sinó també peticions `POST` per crear-les, `DELETE` per eliminar-les, o `PUT` o `PATCH` per modificar-les.

En aquesta unitat estudiarem com fer peticions `GET` i `POST`. A la següent, veurem com fer peticions `DELETE`, `PUT` i `PATCH`.

### Peticions GET

Les peticions `GET` s'utilitzen principalment per sol·licitar dades d'un recurs específic. En el context de Fetch, una petició `GET` es porta a terme de manera senzilla, ja que és el mètode per defecte de la funció `fetch()`.

A l'exemple següent, `fetch()` envia una petició `GET` a la URL proporcionada, que en aquest cas correspon a l'API [PokéApi](https://pokeapi.co) que proporciona informació sobre pokémons. La resposta, que arriba en format JSON, es processa i després s'utilitza com es necessiti, en aquest cas mostrant-la per consola. Aquesta facilitat d'ús i la capacitat de manejar respostes asíncrones de manera eficient fan de Fetch una eina bàsica per bregar amb peticions `GET` en aplicacions web modernes.

In [2]:
fetch('https://pokeapi.co/api/v2/pokemon')
  .then((response) => response.json())               // Converts the response to JSON
  .then((data) => console.log(data))                 // Logs the received data
  .catch((error) => console.error('Error:', error)); // Logs an error if there is one

Promise { <pending> }

{
  count: 1292,
  next: 'https://pokeapi.co/api/v2/pokemon?offset=20&limit=20',
  previous: null,
  results: [
    { name: 'bulbasaur', url: 'https://pokeapi.co/api/v2/pokemon/1/' },
    { name: 'ivysaur', url: 'https://pokeapi.co/api/v2/pokemon/2/' },
    { name: 'venusaur', url: 'https://pokeapi.co/api/v2/pokemon/3/' },
    { name: 'charmander', url: 'https://pokeapi.co/api/v2/pokemon/4/' },
    { name: 'charmeleon', url: 'https://pokeapi.co/api/v2/pokemon/5/' },
    { name: 'charizard', url: 'https://pokeapi.co/api/v2/pokemon/6/' },
    { name: 'squirtle', url: 'https://pokeapi.co/api/v2/pokemon/7/' },
    { name: 'wartortle', url: 'https://pokeapi.co/api/v2/pokemon/8/' },
    { name: 'blastoise', url: 'https://pokeapi.co/api/v2/pokemon/9/' },
    { name: 'caterpie', url: 'https://pokeapi.co/api/v2/pokemon/10/' },
    { name: 'metapod', url: 'https://pokeapi.co/api/v2/pokemon/11/' },
    {
      name: 'butterfree',
      url: 'https://pokeapi.co/api/v2/pokemon/12/'
    },
    { nam

En aquest exemple hem utilitzat la sintaxi més bàsica `fetch(resource)`. Però en realitat la funció `fetch()` admet un segon paràmetre (habitualment anomenat `options`) que consisteix en un objecte de configuració que ens permet especificar el mètode HTTP per utilitzar, els encapçalaments de la petició, el cos de la petició, etc. En el cas del mètode HTTP, el valor per defecte és `GET`, per la qual cosa no cal especificar si volem fer una petició `GET`. Per això el codi de l'exemple anterior funciona bé sense especificar el paràmetre `options`.

Tanmateix, en algunes API caldrà afegir aquest paràmetre per poder especificar capçaleres amb, per exemple, claus d'accés a l'API. A l'exemple següent, mostrem com realitzaríem aquesta petició, amb una API imaginària que podria requerir una clau d'accés a la capçalera `Authorization`:

In [ ]:
async function logMeteoData() {
  const response = await fetch('https://my-meteo-station-cloud.com/api/v4/readings', {
    method: 'GET', // Default value, can be omitted
    headers: {
      'Content-Type': 'application/json',
      Authorization: 'Token c8398a1be479f0e1070bac83aaf6fe6f3a1be471',
    },
  })
  
  if (response.status !== 200) {
    console.warn(`Response status code: ${response.status} - ignoring response`)
  } else {
    const data = await response.json()
    console.log(data)
  }
}

logMeteoData()

Promise { <pending> }

TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11372:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async logMeteoData (evalmachine.<anonymous>:2:20)

> <br>🚨&nbsp; **Atenció**
>
> L'exemple anterior utilitza una API imaginària, per la qual cosa el codi no funcionarà. Però us ha de servir per entendre com s'especifiquen les capçaleres en una petició `GET` amb Fetch. Busqueu alguna API que necessiti una clau d'accés a la capçalera `Authorization`, registreu un compte gratuït per obtenir aquesta clau i proveu de fer una petició `GET` a aquesta API mitjançant Fetch.
> <br><br>

### Peticions POST

Les peticions `POST` s'usen per enviar dades a un servidor i crear un recurs. A diferència de les peticions `GET`, les `POST` poden incloure un cos (body) que contingui la informació que es vol enviar.

Com ja hem vist, el mètode per defecte és `GET`, per la qual cosa si volem fer una petició `POST` serà imprescindible informar al segon paràmetre `options` de la funció `fetch()`. En aquest paràmetre, especificarem el mètode `POST` i el cos de la petició. Vegem-ho amb un primer exemple:

In [3]:
const newPost = {
  "userId": 10,
  "title": "My new awesome post",
  "body": "Some content of the post."
};

fetch('https://jsonplaceholder.typicode.com/posts', {
    method: 'POST',
    headers: {
        'Content-Type': 'application/json'
    },
    body: JSON.stringify(newPost) // Data to be sent to create a new post
})
  .then((response) => {
    console.log(response.status); // Will show you the status code
    response.json()
  })
  .then((data) => console.log('Success:', data))
  .catch((error) => console.error('Error:', error));

Promise { <pending> }

201
Success: undefined


En aquest exemple, enviem una sol·licitud `POST` a un servidor amb dades en format JSON en el cos. És important establir el tipus de contingut a les capçaleres perquè el servidor sàpiga com processar el cos de la sol·licitud. 

Observem que la resposta que obtenim és un objecte de tipus `Response` amb el codi `201` que indica que la petició se ha fet correctament i, per tant, s'ha creat un nou recurs. A més, si inspeccionem la resposta veiem que el cos de la resposta inclou el recurs creat en format JSON. En aquest objecte que rebem com a resposta, podem veure que el nou recurs té un identificador `id` que ha estat assignat automàticament per l'API, ja que no formava part de `newPost`.

> <br>🚨&nbsp; **Atenció**
>
> És molt senzill trobar API públiques que permetin fer peticions `GET`, fins i tot sense necessitat de registrar-se. Tanmateix, és més complicat trobar API públiques que permetin fer peticions `POST`. Això té bastant de sentit si teniu en compte que una API que ens proporcioni informació bursàtil o meteorològica no ens hauria de permetre modificar aquesta informació, sinó només consultar-la. Passa el mateix amb les API de pokémons, personatges de Star Wars o qualsevol altre tipus d'API que ens proporcioni una informació que no té sentit que modifiquem. Especialment perquè si tots els usuaris de l'API poguessin modificar les dades, aquestes no trigarien a quedar completament deteriorades.
>
> No obstant això, per poder practicar amb peticions `POST` (i a la unitat següent, `PUT`, `DELETE` i `PATCH`) observeu que a l'exemple anterior us proposem utilitzar una API pública que sí que ens permet crear i modificar dades. Es tracta de l'API [JSONPlaceholder](https://jsonplaceholder.typicode.com). En concret utilitzarem l'endpoint https://jsonplaceholder.typicode.com/posts, amb el qual simularem la creació i modificació d'articles. Aquesta API ens permet crear i modificar dades de prova, tot i que aquestes dades no s'emmagatzemaran realment a cap lloc.
> <br><br>

Mitjançant una petició `POST` també podem enviar al servidor un fitxer que l'usuari hagi seleccionat des de la seva màquina. Per a això necessitem utilitzar un formulari, creat en el document HTML mitjançant l'etiqueta `<form>`, i un camp de tipus `file` que podem aconseguir amb `<input type="file">`. Vegem-ho amb un exemple en el qual enviem un text i un fitxer:

In [ ]:
const formData = new FormData();
const fileField = document.querySelector("input[type='file']");

formData.append("username", "uoc123");
formData.append("homework", fileField.files[0]);

fetch("https://some-api-accepting-post-requests.com/v2/homework", {
  method: "POST",
  body: formData,
})
  .then((response) => response.json())
  .then((data) => console.log("Success:", data))
  .catch((error) => console.error("Error:", error));

> <br>🚨&nbsp; **Atenció**
>
> L'exemple anterior us donarà error si l'intenteu executar, perquè l'API, com podeu observar, no és real. Però us pot servir per entendre com s'envia un fitxer mitjançant una petició `POST` amb Fetch. Si us heu d'enfrontar a una API que permeti enviar fitxers mitjançant peticions `POST`, proveu de fer una petició amb la sintaxi indicada més amunt.
> <br><br>

> <br>💫&nbsp; **Ampliació**
>
> Tots els tipus de peticions que podem portar a terme amb Fetch admeten múltiples configuracions. Queda fora de l'abast d'aquests materials aprofundir-hi, però si voleu ampliar coneixements podeu consultar la documentació oficial de Fetch a la [MDN](https://developer.mozilla.org/es/docs/Web/API/Fetch_API/Using_Fetch).
> <br><br>